### PR TITLE
Generate Objective-C and C++ protos via CMake

### PR DESCRIPTION
**Pull request by wilhuff**
Adds Objective-C and C++ invocations to `build_protos.py`.

CMake now generates in-place: all generators are idempotent.

Note the diffbase into #1777.

### CODE REVIEW

File: Firestore/Protos/build_protos.py [hunk 1]
```diff
@@ -188,6 +194,36 @@ def __stub_non_buildable_files(self, out_dir):
...
+    self.proto_files = proto_files
+
+  def run(self):
+    out_dir = os.path.join(self.args.output_dir, 'cpp')
+    mkdir(out_dir)
+
+    self.__run_generator(out_dir)
+
+    sources = collect_files(out_dir, '.pb.h', '.pb.cc')
+    # TODO(wilhuff): strip trailing whitespace?
```
Comment (rsgowman) [inline Firestore/Protos/build_protos.py]: It's not something I care about. IMO, the fewer post process steps applied to generated code, the better.

File: Firestore/Protos/CMakeLists.txt [hunk 2]
```diff
@@ -157,6 +157,12 @@ add_custom_command(
     ${PROTOBUF_PYTHON}
 )
 
+add_custom_target(
```
Comment (rsgowman) [inline Firestore/Protos/CMakeLists.txt]: Your other targets have an explicit COMMENT (which makes the output a little nicer). I'd suggest adding one here too.

File: Firestore/Protos/build_protos.py [hunk 3]
```diff
@@ -93,6 +99,10 @@ def main():
   if args.nanopb:
     NanopbGenerator(args, nanopb_proto_files).run()
 
+  proto_files = remove_well_known_protos(nanopb_proto_files)
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/Protos/build_protos.py]: Done: all_proto_files

File: Firestore/Protos/build_protos.py [hunk 4]
```diff
@@ -133,14 +138,88 @@ def __run_generator(self, out_dir):
...
+  We only need them for nanopb.
+
+  Args:
+    filenames: A list of filenames, each naming a .proto file.
+
+  Returns:
+    The filenames with members of google/protobuf removed.
+  """
+
+  return [f for f in filenames if 'protos/google/protobuf/' not in f]
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/Protos/build_protos.py]: Changed in a later commit.

File: Firestore/Protos/CMakeLists.txt (hunk 2 above)
Comment (rsgowman) [reply Firestore/Protos/CMakeLists.txt]: Oh good point; never mind.

... Hm. "Generatoring"?  See line 168.

Comment (wilhuff) [reply Firestore/Protos/CMakeLists.txt]: Generatoring: it's fun!

**Pull request merged**